flocker/control/test/test_persistence.py

```diff
@@ -249,3 +332,193 @@ def test_complex_keys(self):
+        if config['version'] != 2:
+            raise ConfigurationMigrationError(
+                "Supplied configuration was not a valid v2 config."
+            )
+        return json.dumps({"version": 3, "configuration": "fake"})
+
+
+class MigrateConfigurationTests(SynchronousTestCase):
+    """
+    Tests for ``migrate_configuration``.
+    """
+    v1_config = json.dumps({"version": 1})
+
+    def test_error_on_undefined_migration_path(self):
+        """
+        A ``MissingMigrationError`` is raised if a migration path
+        from one version to another cannot be found in the supplied
+        migration class.
+        """
+        e = self.assertRaises(
+            MissingMigrationError,
+            migrate_configuration, 1, 4, self.v1_config, StubMigration
+        )
+        expected_error = (
+            u'Unable to find a migration path for a version 3 to '
+            u'version 4 configuration. No migration method '
+            u'upgrade_from_v3 could be found.'
+        )
+        self.assertEqual(e.message, expected_error)
+
+    def test_sequential_migrations(self):
+        """
+        A migration from one configuration version to another will
+        sequentially perform all necessary upgrades, e.g. v1 to v2 followed
+        by v2 to v3.
+        """
+        # Get a valid v2 config.
+        v2_config = migrate_configuration(1, 2, self.v1_config, StubMigration)
+        # Perform two sequential migrations to get from v1 to v3, starting
+        # with a v1 config.
+        result = migrate_configuration(1, 3, self.v1_config, StubMigration)
+        # Compare the v1 --> v3 upgrade to the direct result of the
+        # v2 --> v3 upgrade on the v2 config, Both should be identical
+        # and valid v3 configs.
+        self.assertEqual(result, StubMigration.upgrade_from_v2(v2_config))
+
+
+def manifestation_generator(random, _):
+    """
+    Hypothesis strategy generator to build a range of manifestations.
+    Generates a mapping of dataset IDs to ``Manifestation`` objects,
+    providing at random between 0 and 5 manifestations, with randomized
+    datasets and attributes.
+    """
+    manifestations = dict()
+    for count in range(0, random.randint(0, 5)):
+        dataset_id = unicode(UUID(int=random.getrandbits(128)))
+        manifestation = Manifestation(
+            primary=bool(random.getrandbits(1)),
+            dataset=Dataset(
+                dataset_id=dataset_id,
+                maximum_size=random.randint(1000000, 100000000)
+            )
+        )
+        manifestations[dataset_id] = manifestation
+    return manifestations
+
+
+ST_MANIFESTATIONS = st.basic(generate=manifestation_generator)
+UUIDS = st.basic(generate=lambda r, _: UUID(int=r.getrandbits(128)))
+ALPHABET = u"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijlmnopqrstuvwxyz"
+SIMPLE_TEXT = st.text(
+    alphabet=ALPHABET, min_size=4, max_size=10, average_size=6
+)
+IMAGES = st.builds(DockerImage, tag=SIMPLE_TEXT, repository=SIMPLE_TEXT)
```

Comment:
I didn't think there was a need to test anything beyond ordinary alphabet characters. They don't technically have to be `string.letters`, the objects can be constructed with any unicode, I've just never encountered a Docker image name that uses a snowman symbol in its name.